### PR TITLE
Make blockchain agent execute transactions on-chain

### PR DESCRIPTION
**Pull request by gabrielfior**
-> Added new tools (write transaction to blockchain)
-> Using Autogen 0.4.X (officially supported by Microsoft + partners, 0.2.X is supported by community)
-> 0.4.X follows the CrewAI style, also adds async functions wider usage

### CODE REVIEW

File: prediction_market_agent/tools/mech/utils.py [hunk 1]
```diff
@@ -43,7 +43,27 @@ class MechTool(str, Enum):
...
+        return True
+    return False
+
+
 def mech_request(question: str, mech_tool: MechTool) -> MechResponse:
+    if not is_package_available("mech_client"):
+        raise EnvironmentError("Package `mech_client` is not installed.")
+
+    from mech_client.interact import ConfirmationType, interact
+
```
Comment (gabrielfior) [inline prediction_market_agent/tools/mech/utils.py:66]: Conditionally importing mech_client if it's in the environment, otherwise throws an error.

File: prediction_market_agent/agents/blockchain_coding_agent/streamlit_console.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    finally:
+        sys.stdout = original_stdout
+
+
+async def streamlit_console(stream: AsyncGenerator[AgentMessage | T, None]) -> None:
+    """
+    Displays stream of messages as Streamlit st.info elements.
+    """
+    with redirect_stdout_to_print():
+        await Console(stream)
```
Comment (gabrielfior) [inline prediction_market_agent/agents/blockchain_coding_agent/streamlit_console.py:35]: This is similar to [Console](https://github.com/microsoft/autogen/blob/1f90dc5ea91a1eae2fce087b4e6feaab87d3dc92/python/packages/autogen-agentchat/src/autogen_agentchat/task/_console.py#L24) but with streamlit's `st.info` for displaying the agent's messages.

File: prediction_market_agent/agents/blockchain_coding_agent/agents.py [hunk 3]
```diff
@@ -1,72 +1,44 @@
...
-import streamlit as st
-from autogen import Agent, ConversableAgent
-from autogen.coding.jupyter import JupyterCodeExecutor, LocalJupyterServer
-
-from prediction_market_agent.agents.blockchain_coding_agent.prompts import (
-    code_writer_system_message,
+from autogen_agentchat.agents import AssistantAgent, BaseChatAgent
+from autogen_agentchat.task import TextMentionTermination
+from autogen_agentchat.teams import BaseGroupChat, RoundRobinGroupChat
+from autogen_ext.models._openai._openai_client import OpenAIChatCompletionClient
```
[the remark this answers is not in the document]
Comment (kongzii) [reply prediction_market_agent/agents/blockchain_coding_agent/agents.py:4]: Agreed with rabbit, isn't `from autogen_ext.models import OpenAIChatCompletionClient` working? Or maybe from some other place?

File: prediction_market_agent/agents/blockchain_coding_agent/app.py [hunk 4]
```diff
@@ -1,51 +1,57 @@
...
 import streamlit as st
-from autogen import Cache
 
-from prediction_market_agent.agents.blockchain_coding_agent.agents import (
-    get_code_executor_agent,
-    get_code_writer_agent,
+from prediction_market_agent.agents.blockchain_coding_agent.agents import get_agent_team
+from prediction_market_agent.agents.blockchain_coding_agent.prompts import (
+    message_read,
+    message_write,
```
Comment (kongzii) [inline prediction_market_agent/agents/blockchain_coding_agent/app.py:9]: Keep the practice of  global variables being upper case please

File: prediction_market_agent/agents/blockchain_coding_agent/app.py [hunk 5]
```diff
@@ -1,51 +1,57 @@
...
-from prediction_market_agent.agents.blockchain_coding_agent.functions import (
-    register_all_functions,
+from prediction_market_agent.agents.blockchain_coding_agent.streamlit_console import (
+    streamlit_console,
 )
 
-st.write("""# AutoGen Chat Agents""")
+
+def is_user_prompt_not_null(prompt: str) -> bool:
+    return prompt is not None and prompt != ""
```
Comment (kongzii) [inline prediction_market_agent/agents/blockchain_coding_agent/app.py:17]:
```suggestion
    return prompt != ""
```

File: prediction_market_agent/agents/blockchain_coding_agent/app.py [hunk 6]
```diff
@@ -1,51 +1,57 @@
...
+    2. Load one of the predefined prompts (Read, Write).
+    """
     )
-    user_input = st.chat_input(
-        placeholder="Copy prompt from above",
+    st.divider()
+    user_input = st.text_input(label="Enter a new prompt if so desired", key="prompt")
+    prompt_option = st.selectbox(
+        "Select a prompt option",
+        (UserPrompt.READ_CONDITIONAL_TOKENS_BALANCE, UserPrompt.WRITE_USDC_APPROVAL),
```
Comment (kongzii) [inline prediction_market_agent/agents/blockchain_coding_agent/app.py:43]:
```suggestion
        list(UserPrompt),
```

will just work and also be automatically updated in the future

File: prediction_market_agent/agents/blockchain_coding_agent/app.py [hunk 7]
```diff
@@ -1,51 +1,57 @@
...
-                )
-                # The correct result should be 289421015806737773 (as of block 37141392)
-
-        loop.run_until_complete(initiate_chat())
+
+    submit_button = st.button("Submit", key="button")
+
+    if submit_button:
+        message = st.session_state.prompt_option
+        if st.session_state.prompt is not None and st.session_state.prompt != "":
```
Comment (kongzii) [inline prediction_market_agent/agents/blockchain_coding_agent/app.py:54]:
```suggestion
        if is_user_prompt_not_null(st.session_state.prompt):
```

File: prediction_market_agent/agents/blockchain_coding_agent/functions.py [hunk 8]
```diff
@@ -25,47 +28,44 @@ def tavily_search(
...
-    w3 = OmenConditionalTokenContract().get_web3()
-    if not is_contract(w3, Web3.to_checksum_address(contract_address)):
-        raise ValueError(f"{contract_address=} is not a contract on Gnosis Chain.")
-    read_not_proxy = fetch_read_methods(contract_address)
-    read_proxy = fetch_read_methods_proxy(contract_address)
-    # ToDo - Fetch write methods
-    return read_not_proxy + read_proxy
+def get_rpc_endpoint() -> str:
+    # return "https://rpc.gnosischain.com"
+    return "http://localhost:8545"
```
Comment (kongzii) [inline prediction_market_agent/agents/blockchain_coding_agent/functions.py:33]: Why did you change this back to a hard-coded string?

File: prediction_market_agent/agents/blockchain_coding_agent/functions.py [hunk 9]
```diff
@@ -25,47 +28,44 @@ def tavily_search(
...
-    return data
-
-
-def get_rpc_endpoint() -> str:
-    return RPCConfig().gnosis_rpc_url
-
-
-def checksum_address(address: str) -> ChecksumAddress:
-    return Web3.to_checksum_address(address)
+def fetch_source_code_and_abi_from_contract(
```
Comment (kongzii) [inline prediction_market_agent/agents/blockchain_coding_agent/functions.py:44]: Any reason to not re-use `persistent_inmemory_cache` that we already have?

```suggestion
@persistent_inmemory_cache
def fetch_source_code_and_abi_from_contract(
```

Then we don't need a new caching logic with `CachedSession`

File: prediction_market_agent/agents/blockchain_coding_agent/functions.py [hunk 10]
```diff
@@ -75,62 +75,45 @@ def execute_read_function(
...
-        caller=caller_agent,  # The assistant agent can suggest calls to the calculator.
-        executor=executor_agent,  # The user proxy agent can execute the calculator calls.
-        name="fetch_read_methods_from_blockscout",  # By default, the function name is used as the tool name.
-        description="Function for fetching the ABI from a verified smart contract on Gnosis Chain",
-        # A description of the tool.
-    )
+    output = contract.functions[function_name](
+        *parse_function_params(function_params)
+    ).call()  # type: ignore # TODO: Fix Mypy, as this works just OK.
+    return output
```
Comment (kongzii) [inline prediction_market_agent/agents/blockchain_coding_agent/functions.py:93]: Why didn't you use `call_function_on_contract` inside of this function instead of copy-pasting the code from PMAT?

File: prediction_market_agent/agents/blockchain_coding_agent/functions.py [hunk 11]
```diff
@@ -75,62 +75,45 @@ def execute_read_function(
...
+##### Write
 
-    register_function(
-        checksum_address,
-        caller=caller_agent,
-        executor=executor_agent,
-        name="checksum_address",
-        description="Extracts the checksummed address of an address.",
+
+def get_private_key() -> PrivateKey:
```
Comment (kongzii) [inline prediction_market_agent/agents/blockchain_coding_agent/functions.py:99]: What's the reason for giving an agent the ability to read its private key? That's not very safe.

Or, if `get_private_key` isn't registered as agent's function, then you can use `APIKeys().bet_from_private_key` directly?

File: prediction_market_agent/agents/blockchain_coding_agent/functions.py [hunk 12]
```diff
@@ -75,62 +75,45 @@ def execute_read_function(
...
+    return APIKeys().bet_from_private_key
+
+
+def execute_write_function(
+    contract_address: str,
+    abi: str,
+    function_name: str,
+    function_params: t.List[t.Any],
+) -> TxReceipt:
+    # ToDo - Do not use PMAT if possible.
```
Comment (kongzii) [inline prediction_market_agent/agents/blockchain_coding_agent/functions.py:109]: Why not? If there is something implemented in PMAT already, I don't see reason why not just wrap it into autogen-compatible functions?

File: prediction_market_agent/agents/blockchain_coding_agent/run_agent04_read.py [hunk 13]
```diff
@@ -0,0 +1,15 @@
+import asyncio
```
Comment (kongzii) [inline prediction_market_agent/agents/blockchain_coding_agent/run_agent04_read.py:1]: Can you create a scripts directory like `prediction_market_agent/agents/blockchain_coding_agent/scripts` and move these run files there? Just to keep clean structure about what is a script and what not

File: prediction_market_agent/agents/blockchain_coding_agent/functions.py (hunk 8 above)
Comment (gabrielfior) [reply prediction_market_agent/agents/blockchain_coding_agent/functions.py:33]: My mistake. Reverting and closing this.

File: prediction_market_agent/agents/blockchain_coding_agent/functions.py (hunk 10 above)
Comment (gabrielfior) [reply prediction_market_agent/agents/blockchain_coding_agent/functions.py:93]: My rationale was not to use PMAT to allow the agent to rely purely on web3py (I was considering the case where the agent also writes some code).
However, here it's purely doing function calling, so sure, let's use PMAT for simplified logic.
Implement this also for the other comment.

File: prediction_market_agent/agents/blockchain_coding_agent/functions.py (hunk 12 above)
Comment (gabrielfior) [reply prediction_market_agent/agents/blockchain_coding_agent/functions.py:109]: See https://github.com/gnosis/prediction-market-agent/pull/566/files#r1872193547
Closing this since same idea from other comment

File: prediction_market_agent/agents/blockchain_coding_agent/functions.py (hunk 11 above)
Comment (gabrielfior) [reply prediction_market_agent/agents/blockchain_coding_agent/functions.py:100]: Agree. Removed and using now `APIKeys().private_key` directly

File: prediction_market_agent/agents/blockchain_coding_agent/agents.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (gabrielfior) [reply prediction_market_agent/agents/blockchain_coding_agent/agents.py:4]: I disagree with rabbit, the correct import is
```
from autogen_ext.models import OpenAIChatCompletionClient
```
as exemplified in https://microsoft.github.io/autogen/dev/user-guide/agentchat-user-guide/quickstart.html#quickstart

File: prediction_market_agent/agents/blockchain_coding_agent/app.py (hunk 4 above)
Comment (gabrielfior) [reply prediction_market_agent/agents/blockchain_coding_agent/app.py]: Renamed

File: prediction_market_agent/agents/blockchain_coding_agent/functions.py [hunk 14]
```diff
@@ -25,47 +28,44 @@
...
+    r.raise_for_status()
+    data_parsed = SmartContractResponse.model_validate(r.json())
+    source_code = data_parsed.source_code
+    abi = data_parsed.abi
+    # If proxy, expand abi and source code with implementation data.
+    if data_parsed.implementations:
+        for proxy in data_parsed.implementations:
+            proxy_container = fetch_source_code_and_abi_from_contract(proxy.address)
+            abi.extend(proxy_container.abi)
+            source_code += "\n" + proxy_container.source_code
```
[the remark this answers is not in the document]
Comment (gabrielfior) [reply prediction_market_agent/agents/blockchain_coding_agent/functions.py:60]: Good catch rabbit. Implemented.

**Review by kongzii — APPROVED** [review]
